query-planner-js/src/composedSchema/buildComposedSchema.ts

```diff
@@ -0,0 +1,240 @@
+  FederationTypeMetadata,
+  FieldSet,
+  GraphMap,
+} from './metadata';
+
+export function buildComposedSchema(document: DocumentNode): GraphQLSchema {
+  const schema = buildASTSchema(document);
+
+  // TODO: We should follow the Bootstrap algorithm from the Core Schema spec
+  // to handle renames of @core itself.
+  const coreName = 'core';
+
+  const coreDirective = schema.getDirective(coreName);
+
+  assert(coreDirective, `Expected core schema, but can't find @core directive`);
+
+  // TODO: We should follow the CollectFeatures algorithm from the Core Schema
+  // spec here, and use the collected features to validate feature
+  // versions and handle renames.
+
+  const coreDirectivesArgs = getArgumentValuesForRepeatableDirective(
+    coreDirective,
+    schema.astNode!,
+  );
+
+  for (const coreDirectiveArgs of coreDirectivesArgs) {
+    const feature: string = coreDirectiveArgs['feature'];
+
+    if (
+      !(
+        feature === 'https://lib.apollo.dev/core/v0.1' ||
+        feature === 'https://lib.apollo.dev/join/v0.1'
+      )
+    ) {
+      throw new GraphQLError(
+        `Unsupported core schema feature and/or version: ${feature}`,
+        schema.astNode!,
+      );
+    }
+  }
+
+  const joinName = 'join';
+
+  function getJoinDirective(name: string) {
+    const fullyQualifiedName = `${joinName}__${name}`;
+
+    const directive = schema.getDirective(fullyQualifiedName);
+    assert(
+      directive,
+      `Composed schema should define @${fullyQualifiedName} directive`,
+    );
+    return directive;
+  }
+
+  const ownerDirective = getJoinDirective('owner');
+  const typeDirective = getJoinDirective('type');
+  const fieldDirective = getJoinDirective('field');
+  const graphDirective = getJoinDirective('graph');
+
+  const graphEnumType = schema.getType(`${joinName}__Graph`);
+  assert(isEnumType(graphEnumType));
+
+  const graphMap: GraphMap = Object.create(null);
+
+  schema.extensions = {
+    ...schema.extensions,
+    federation: {
+      graphs: graphMap,
+    },
+  };
+
+  for (const graphValue of graphEnumType.getValues()) {
+    const name = graphValue.name;
+
+    const graphDirectiveArgs = getArgumentValuesForDirective(
+      graphDirective,
+      graphValue.astNode!,
+    );
+    assert(
+      graphDirectiveArgs,
+      `${graphEnumType.name} value ${name} in composed schema should have a @${graphDirective.name} directive`,
+    );
+
+    const serviceName: string = graphDirectiveArgs['name'];
+    const url: string = graphDirectiveArgs['url'];
+
+    graphMap[name] = {
+      serviceName,
+      url,
+    };
+  }
+
+  for (const type of Object.values(schema.getTypeMap())) {
+    if (isIntrospectionType(type)) continue;
+
+    // We currently only allow join spec directives on object types.
+    if (!isObjectType(type)) continue;
+
+    assert(
+      type.astNode,
+      `GraphQL type "${type.name}" should contain AST nodes`,
+    );
+
+    const ownerDirectiveArgs = getArgumentValuesForDirective(
```

Comment:
I dunno if feedback on the actual directives is helpful now, but I don't get why `@join__owner` is its own directive rather than just an `owner: Boolean` argument to `@join__type` that you can only set on at most one directive.

Reply:
I think one benefit of having it as a separate directive is that validation will catch putting more than one (non-repeatable) `@join__owner` on a type.